File: typescript/src/molkky.ts
export class Molkky {
  private currentScore: number = 0;
  private fails: number = 0;
  private overFlow = false;
  private currentState: string | null = null;
  private running = true;
  private duplicate = false;

  public score() {
    return this.currentScore;
  }

  public shoot(pinValue: number[]) {
    if (this.currentState == "WON") {
      this.currentState = "GAME ALREADY WON";
    }
    if (!this.running && this.fails == 3) {
      this.currentState = "GAME ALREADY LOST";
    } else {
      if (pinValue.length >= 1 && this.running) {
        let tmpPins: number[] = [];
        for (const pin of pinValue) {
          if (!tmpPins.find((v) => v == pin) && pin < 13 && pin >= 1) {
            tmpPins.push(pin);
          }
        }
        if (tmpPins.length != pinValue.length) {
          this.duplicate = true;
        }
        if (this.currentScore + tmpPins.length > 50) {
          this.overFlow = true;
          this.currentScore = 25;
        } else {
          this.overFlow = false;
          this.currentScore += this.duplicate
            ? tmpPins.length
            : pinValue.length;
        }
        this.fails -= 1;
        if (pinValue.length == 1) {
          if (this.currentScore + pinValue[0] - pinValue.length < 51) {
            this.currentScore +=
              pinValue[0] > 0 && 13 > pinValue[0] ? pinValue[0] - 1 : 0;
            this.overFlow = false;
          } else {
            this.currentScore = 25;
            this.overFlow = true;
          }
          if (this.overFlow) {
            this.currentState = "SCORE OVERFLOW";
          }
        }
        if (tmpPins.length == 1 && this.duplicate) {
          if (50 > this.currentScore + tmpPins[0] - (tmpPins.length + 1)) {
            this.currentScore +=
              tmpPins[0] < 1 || tmpPins[0] > 12 ? -1 : tmpPins[0] - 1;
            var overFlow = false;
          } else {
            this.currentScore = 25;
            overFlow = true;
          }
          if (overFlow) {
            this.currentState = "SCORE OVERFLOW";
          }
        }
        if (this.currentScore == 50) {
          this.currentState = "WON";
          this.running = false;
        }
      } else {
        if (this.fails < 1) {
          this.fails = 0;
        }
        if (this.fails > 1) {
          this.currentState = "LOST";
          this.running = false;
        }
        this.fails += 1;
      }
    }
  }

  public state(): string {
    if (this.currentState == null) {
      this.currentState = "RUNNING";
    }

    return this.currentState;
  }
}

File: python/molkky.py
class Molkky:
    def __init__(self):
        self.current_score = 0
        self.fails = 0
        self.over_flow = False
        self.current_state = None
        self.running = True
        self.duplicate = False

    def score(self):
        return self.current_score

    def shoot(self, pin_value):
        if self.current_state == "WON":
            self.current_state = "GAME ALREADY WON"

        if not self.running and self.fails == 3:
            self.current_state = "GAME ALREADY LOST"
        else:
            if len(pin_value) >= 1 and self.running:
                tmp_pins = []
                for pin in pin_value:
                    if pin not in tmp_pins and 1 <= pin < 13:
                        tmp_pins.append(pin)

                if len(tmp_pins) != len(pin_value):
                    self.duplicate = True

                if self.current_score + len(tmp_pins) > 50:
                    self.over_flow = True
                    self.current_score = 25
                else:
                    self.over_flow = False
                    self.current_score += len(tmp_pins) if self.duplicate else len(pin_value)

                self.fails -= 1

                if len(pin_value) == 1:
                    if self.current_score + pin_value[0] - len(pin_value) < 51:
                        self.current_score += pin_value[0] - 1 if 0 < pin_value[0] < 13 else 0
                        self.over_flow = False
                    else:
                        self.current_score = 25
                        self.over_flow = True

                    if self.over_flow:
                        self.current_state = "SCORE OVERFLOW"

                if len(tmp_pins) == 1 and self.duplicate:
                    if 50 > self.current_score + tmp_pins[0] - (len(tmp_pins) + 1):
                        self.current_score += -1 if tmp_pins[0] < 1 or tmp_pins[0] > 12 else tmp_pins[0] - 1
                        local_over_flow = False
                    else:
                        self.current_score = 25
                        local_over_flow = True

                    if local_over_flow:
                        self.current_state = "SCORE OVERFLOW"

                if self.current_score == 50:
                    self.current_state = "WON"
                    self.running = False
            else:
                if self.fails < 1:
                    self.fails = 0
                if self.fails > 1:
                    self.current_state = "LOST"
                    self.running = False
                self.fails += 1

    def state(self):
        if self.current_state is None or self.current_state == "SCORE OVERFLOW":
            self.current_state = "RUNNING"

        return self.current_state
